### PR TITLE
Web html emoji issule fix

### PR DESCRIPTION
**Pull request by zjrwdl**
Emoji display problem in flutter web html render #131337

before this pr:
image

after this pr:
image

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/131337

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by gildaswise** [comment]
Hey, is this fix still planned? I'm also having issues with the html renderer with other ZWJ emojis 

image

**Comment by harryterkelsen** [comment]
https://github.com/flutter/engine/pull/44247#issuecomment-1661950687

Your PR affects text layout, and the failing tests are related to text layout. Please resolve the failing tests. Also, please add a new test which will fail before your change and which succeeds after your change.

**Comment by goderbauer** [comment]
(triage) @zjrwdl Just checking in to see if you still have plans to address the feedback given above?

File: lib/web_ui/lib/src/engine/text/canvas_paragraph.dart [hunk 1]
```diff
@@ -169,15 +170,18 @@ class CanvasParagraph implements ui.Paragraph {
...
           spanElement.setAttribute('dir', 'rtl');
         }
-        applyTextStyleToElement(element: spanElement, style: fragment.style);
-        _positionSpanElement(spanElement, line, fragment);
-
-        spanElement.appendText(text);
+        applyTextStyleToElement(
+            element: spanElement, style: firstFragment.style);
+        _positionSpanElement(spanElement, line, firstFragment);
+        spanElement.appendText(fullText);
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text/canvas_paragraph.dart:184]: It looks like you are putting the entire line into a single `<span>` element, is that correct?

This won't work. One line of text could contain fragments with different styles. You can't just put all of them in the same `<span>` with the same style.

**Review by mdebbar — CHANGES_REQUESTED** [review]

**Comment by goderbauer** [comment]
(triage) @zjrwdl Do you still have plans to come back and address the feedback given above? Thank you.

**Comment by goderbauer** [comment]
I am going to close this one since there was no follow-up. Feel free to reopen if you find the time to work on this again. Thank you.

**Pull request closed without merging**